### PR TITLE
Add ClickHouse adapter

### PR DESCRIPTION
**Pull request by Felixoid**
Hello. #77 has implemented a workaround for using ClickHouse via MySQL client. This PR implements adapter to use the native client for interactive and non-interactive mode.

### CODE REVIEW

**Comment by tpope** [comment]
Lets start by making this `clickhouse-client` only, as `curl` opens up a can of worms that I think would be easier to address in a followup PR.

Is your use of `secure` with `0` as disabled based on anything in particular? [The documentation](https://clickhouse.tech/docs/en/interfaces/cli/#configuration_files) suggests that `True` and `False` are the canonical values, so we should probably try to match those or at the very least accept them. In the [MongoDB adapter](https://github.com/kristijanhusak/vim-dadbod/blob/bb76942b178e04768bbe86ee9f998e8795fa5520/autoload/db/adapter/mongodb.vim#L20) we take anything starting with `1` or `t` as true, which I think is a reasonable way to "do what I mean" it. But add a `T` because it's case sensitive.

**Comment by Felixoid** [comment]
Hey Tim, thank you for the review.

> Lets start by making this `clickhouse-client` only, as `curl` opens up a can of worms that I think would be easier to address in a followup PR.

Ok. I get so far, you aren't against the curl implementation for later, but in this PR would like to see a native `clickhouse-client` only implementation. Right? NP, sounds good to me.

> The documentation suggests that True and False are the canonical values

It's a close, but not a complete example. In CH configs, many things are valid `Boolean` values, here's the example (9440 is SSL, 9000 is plain TCP)

```
┌[12:15:51] felixoid@ig-fel:~
└>±> grep 'secure>' /etc/clickhouse-client/config.xml && clickhouse-client 
    <secure>1</secure>
ClickHouse client version 20.10.2.20 (official build).
Connecting to localhost:9440 as user default.
Code: 210. DB::NetException: Connection refused (localhost:9440)

┌[12:16:03] felixoid@ig-fel:~
└>±> grep 'secure>' /etc/clickhouse-client/config.xml && clickhouse-client
    <secure>true</secure>
ClickHouse client version 20.10.2.20 (official build).
Connecting to localhost:9440 as user default.
Code: 210. DB::NetException: Connection refused (localhost:9440)

┌[12:16:15] felixoid@ig-fel:~
└>±> grep 'secure>' /etc/clickhouse-client/config.xml && clickhouse-client
    <secure>TRUE</secure>
ClickHouse client version 20.10.2.20 (official build).
Connecting to localhost:9440 as user default.
Code: 210. DB::NetException: Connection refused (localhost:9440)

┌[12:16:47] felixoid@ig-fel:~
└>±> grep 'secure>' /etc/clickhouse-client/config.xml ; clickhouse-client
ClickHouse client version 20.10.2.20 (official build).
Connecting to localhost:9000 as user default.
Connected to ClickHouse server version 20.10.2 revision 54441.

ig-fel :) Bye.
┌[12:20:35] felixoid@ig-fel:~
└>±> grep 'secure>' /etc/clickhouse-client/config.xml && clickhouse-client
    <secure>0</secure>
ClickHouse client version 20.10.2.20 (official build).
Connecting to localhost:9000 as user default.
Connected to ClickHouse server version 20.10.2 revision 54441.

ig-fel :) Bye.
```

I will reuse the mongodb adapter snippet then. It reflects the logic pretty well :+1:

**Comment by Felixoid** [comment]
And I've found some bug in curl implementation, so, it is indeed a good idea to implement it in the different patch

**Comment by Felixoid** [comment]
Hey, @tpope, it's ready for another review round

File: autoload/db/adapter/clickhouse.vim [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+function! db#adapter#clickhouse#complete_opaque(_) abort
+  return db#adapter#clickhouse#complete_database('clickhouse:///')
+endfunction
+
+function! db#adapter#clickhouse#auth_input() abort
+  return 'SELECT 1'
+endfunction
+
+function! db#adapter#clickhouse#auth_pattern() abort
+  return '^Code: 516. DB::Exception'
```
Comment (tpope) [inline autoload/db/adapter/clickhouse.vim:55]: Let's make that `.` a `\.`, just to more clearly communicate intent.

File: autoload/db/adapter/clickhouse.vim [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+  for [k, v] in items(url.params)
+    let cmd .= ' --' . k . ' ' . shellescape(v)
+  endfor
+  return cmd .
+        \ db#url#as_args(a:url, '--host ', '--port ', '', '--user ', '--port ', '--database ')
+endfunction
+
+function! db#adapter#clickhouse#filter(url) abort
+  return db#adapter#clickhouse#interactive(a:url)
+endfunction
```
Comment (tpope) [inline autoload/db/adapter/clickhouse.vim:44]: You can delete this, `#filter()` falls back to `#interactive()` by default.

**Review by tpope — CHANGES_REQUESTED** [review]

File: autoload/db/adapter/clickhouse.vim (hunk 1 above)
Comment (Felixoid) [reply autoload/db/adapter/clickhouse.vim]: It's fair

**Pull request merged**